### PR TITLE
Rebindable keybindings

### PR DESCRIPTION
**Pull request by ThatLing**
## It's 2018 and you still can't rebind your keybindings? 
I originally made this for Yogstation where it's been running for a few weeks now without any major problems. You can only assign an action to 1 key at a time and you can change your bindings in-game without having to relog. Default mode has no rebindable keybindings but the original controls are still supported.

The available keys are in a big list which is really stupid but I don't know how else to do it. I removed the buttons that allowed mins to change the movement keys since it would be too much effort to fix such a niche feature. GUI looks pretty bad because I'm bad at GUI. As usual, point out any bad code.

:cl:
add: You can now rebind your keybindings in the preferences window.
del: Removed admin buttons for changing movement keys of clients.
/:cl:

### CODE REVIEW

**Comment by optimumtact** [comment]
problem

goonchat has some random ass keybinds hardcoded in to it

File: code/__DEFINES/keybindings.dm [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+	"Numpad9",
+	",",
+	".",
+	"&",
+	"é",
+	"\"",
+	"'",
+	"(",
+	"§",
+	"è",
```
Comment (optimumtact) [inline code/__DEFINES/keybindings.dm]: what the

**Review by optimumtact — CHANGES_REQUESTED** [review]
[12:04] Melon: @oranges hi sorry for bother, is there a known issue where no keybinds are accepted anymore and SPACE does what TAB normally down in default mode?

File: code/__DEFINES/keybindings.dm (hunk 1 above)
Comment (ThatLing) [reply code/__DEFINES/keybindings.dm]: French layout uses these. Like stated in OP, this is really ugly. Any ideas on how to clean this up?

File: code/controllers/subsystem/input.dm [hunk 2]
```diff
@@ -5,22 +5,19 @@ SUBSYSTEM_DEF(input)
 	flags = SS_TICKER
 	priority = FIRE_PRIORITY_INPUT
 	runlevels = RUNLEVELS_DEFAULT | RUNLEVEL_LOBBY
-
+
```
Comment (optimumtact) [inline code/controllers/subsystem/input.dm]: police the stray spacing please

File: code/modules/keybindings/bindings_client.dm [hunk 3]
```diff
@@ -4,50 +4,70 @@
...
 // Called every game tick
 /client/keyLoop()
 	if(holder)
 		holder.keyLoop(src)
 	if(mob.focus)
-		mob.focus.keyLoop(src)
\ No newline at end of file
+		mob.focus.keyLoop(src)
+
+/client/proc/get_ooc()
```
Comment (optimumtact) [inline code/modules/keybindings/bindings_client.dm]: why is this here?

File: code/modules/admin/secrets.dm [hunk 4]
```diff
@@ -590,60 +590,6 @@
 				purrbation.")
 			log_admin("[key_name(usr)] has removed everyone from purrbation.")
 
-		if("flipmovement")
-			if(!check_rights(R_FUN))
-				return
-			if(alert("Flip all movement controls?","Confirm","Yes","Cancel") == "Cancel")
```
Comment (MrStonedOne) [inline code/modules/admin/secrets.dm]: Undocumented change.

Why are you removing these admin buttons?

[`You are going to be expected to document all your changes in the pull request. Failing to do so will mean delaying it as we will have to question why you made the change`](https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#pull-request-process)

Please mention this change in the PR body, (as well as explain how its related to this pull request). Thank you

**Review by MrsTonedOne — CHANGES_REQUESTED** [review]

File: code/modules/admin/secrets.dm (hunk 4 above)
Comment (ThatLing) [reply code/modules/admin/secrets.dm]: Sorry for that, updated OP to include the removal. I removed it because I've never seen anyone use this on Yogstation and I couldn't be arsed to spend much time properly fixing such a niche button.

File: code/modules/client/preferences.dm [hunk 5]
```diff
@@ -588,6 +590,68 @@ GLOBAL_LIST_EMPTY(preferences_datums)
...
+			if(hotkeys)
+				var/button
+				var/button_bound
+
+				dat += "<table><tr><td width='340px' height='300px' valign='top'>"
+				dat += "<h2>Client</h2>"
+				BUTTON_KEY_MOVEMENT("Move North (up)", ACTION_MOVENORTH, NORTH)
+				BUTTON_KEY_MOVEMENT("Move West (left)", ACTION_MOVEWEST, WEST)
+				BUTTON_KEY_MOVEMENT("Move South (down)", ACTION_MOVESOUTH, SOUTH)
+				BUTTON_KEY_MOVEMENT("Move East (right)", ACTION_MOVEEAST, EAST)
```
Comment (ninjanomnom) [inline code/modules/client/preferences.dm]: The actions available per type should be determined by the code not changed manually here.

File: code/modules/keybindings/bindings_client.dm [hunk 6]
```diff
@@ -4,50 +4,70 @@
...
 			return
-		if("F12") // Toggles minimal HUD
+		if(ACTION_MINHUD) // Toggles minimal HUD
 			mob.button_pressed_F12()
 			return
+		if(ACTION_OOC)
+			get_ooc()
+			return
+			
+	var/datum/keyinfo/I = bind.to_keyinfo(_key, A)
```
Comment (ninjanomnom) [inline code/modules/keybindings/bindings_client.dm]: Use a better name here than `I` please

File: code/modules/keybindings/bindings_client.dm [hunk 7]
```diff
@@ -4,50 +4,70 @@
...
+	bind.set_key_up(_key)
+	var/movement = bind.get_movement_dir(_key)
 	if(!(next_move_dir_add & movement))
 		next_move_dir_sub |= movement
 
+	var/A = bind.get_key_action(_key)
+	if(!A)
+		return
+
+	var/datum/keyinfo/I = bind.to_keyinfo(_key, A)
```
Comment (ninjanomnom) [inline code/modules/keybindings/bindings_client.dm]: Same as above

File: code/modules/keybindings/bindings_mob.dm [hunk 8]
```diff
@@ -2,64 +2,76 @@
 // All it takes is one badmin setting their focus to someone else's client to mess things up
 // Or we can have NPC's send actual keypresses and detect that by seeing no client
 
-/mob/key_down(_key, client/user)
-	switch(_key)
-		if("Delete", "H")
+/mob/key_down(datum/keyinfo/I, client/user)
+	switch(I.action)
+		if(ACTION_SAY)
```
Comment (ninjanomnom) [inline code/modules/keybindings/bindings_mob.dm]: The keybindings datum should be handling the actions rather than here.

File: code/controllers/subsystem/input.dm [hunk 9]
```diff
@@ -7,33 +7,27 @@ SUBSYSTEM_DEF(input)
...
 		macro_sets = default_macro_sets
 		return
 
 	default_macro_sets = list(
 		"default" = list(
-			"Tab" = "\".winset \\\"input.focus=true?map.focus=true input.background-color=[COLOR_INPUT_DISABLED]:input.focus=true input.background-color=[COLOR_INPUT_ENABLED]\\\"\"",
-			"O" = "ooc",
-			"T" = "say",
-			"M" = "me",
+			"Tab" = "\".winset \\\"mainwindow.macro=old_default input.focus=true input.background-color=[COLOR_INPUT_ENABLED]\\\"\"",
```
Comment (ninjanomnom) [inline code/controllers/subsystem/input.dm]: This is improperly making it so tab changes your hotkey mode to the old modes.

File: code/modules/admin/secrets.dm (hunk 4 above)
Comment (DaxDupont) [reply code/modules/admin/secrets.dm]: Don't some chemicals or situations other that admin buttons cause movement flipping?

There's something that does it, I just don't recall what.

**Review by ninjanomnom — CHANGES_REQUESTED** [review]
This is a very poor implementation of rebindable controls and I'd prefer it just not be done at all with this implementation.

**Comment by ninjanomnom** [comment]
This needs a lot of work, message me on irc or discord or something if you need help with that.
